tests/controllers/product.test.js

```diff
@@ -113,80 +123,222 @@ tape('POST /products with invalid product data', t => {
+    dropCollectionsAndEnd([Product, User, Token], t)
+  })
+  .catch(err => t.end(err))
 })
 
 // Tests for: DELETE /products/:id
 tape('DELETE /products/:id with id of something not in the database', t => {
-  supertest(server)
+  Promise.all([
+    Product.create(validProduct1),
+    User.create(superUser),
+    makeLoggedInToken(user)
+  ])
+  .then(([product, userSuper, token]) => {
+    return supertest(server)
     .delete('/products/507f1f77bcf86cd799439011')
-    .expect(400)
+    .set('Cookie', `token=${token}`)
+    .expect(404)
     .expect('Content-Type', /json/)
-    .end((err, res) => {
-      if (err) t.fail(err)
-      t.ok(res.body.message, 'Message sent back')
-      t.equal(res.body.message, 'Cannot find document to delete', 'Correct message is sent back')
-      t.end()
-    })
+  })
+  .then(res => {
+    t.ok(res.body.message, 'Message sent back')
+    t.equal(res.body.message, 'No document matching that id', 'Correct message is sent back')
+    dropCollectionsAndEnd([Product, User], t)
+  })
+  .catch(err => t.end(err))
 })
 
 tape('DELETE /products/:id with good ID', t => {
-  Product.create(validProduct1, validProduct2)
-    .then(productToBeDeleted => {
-      supertest(server)
-        .delete(`/products/${productToBeDeleted.id}`)
-        .expect(204)
-        .end((err, res) => {
-          if (err) t.fail(err)
-          t.deepEqual(res.body, {}, 'Nothing returned after deletion')
-          // check our database now has one fewer product
-          Product.find()
-            .then(productsAfterDeletion => {
-              t.equal(productsAfterDeletion.length, 1, 'Products should now be length 1')
-              t.ok(productsAfterDeletion[0].id !== productToBeDeleted.id, 'deleted product should no longer be in the database')
-              dropCollectionAndEnd(Product, t)
-            })
-            .catch(err => {
-              t.fail(err)
-              dropCollectionAndEnd(Product, t)
-            })
-        })
+  Promise.all([
+    Product.create(validProduct1),
+    Product.create(validProduct2),
+    User.create(superUser),
+    makeLoggedInToken(superUser)
+  ])
+  .then(([product, productToBeDeleted, userSuper, token]) => {
+    supertest(server)
+    .delete(`/products/${productToBeDeleted.id}`)
+    .set('Cookie', `token=${token}`)
+    .expect(204)
+    .then(res => {
+      t.deepEqual(res.body, {}, 'Nothing returned after deletion')
+      // check our database now has one fewer product
+      Product.find()
+      .then(productsAfterDeletion => {
+        t.equal(productsAfterDeletion.length, 1, 'Products should now be length 1')
+        t.ok(productsAfterDeletion[0].id !== productToBeDeleted.id, 'deleted product should no longer be in the database')
+        dropCollectionsAndEnd([Product, User], t)
+      })
+      .catch(err => {
+        t.fail(err)
+        dropCollectionsAndEnd([Product, User], t)
+      })
+    })
+    .catch(err => t.end(err))
+  })
+  .catch(err => t.end(err))
```

Comment:
you could flatten all this to just one `catch`, it looks like all you need is the `productToBeDeleted` variable to be defined in the outermost scope

Reply:
yeah so it got complicated when accessing the id to compare to or something, I need to think it through more - I tried manually setting the id in the fixtures file but that threw a load of errors for some reason